### PR TITLE
Implement and test basic Zarr v3

### PR DESCRIPTION
**Pull request by aliddell**
Data generated in `write-zarr-v3-raw.cpp` and `write-zarr-v3-compressed.cpp` were tested as readable using zarr-python.

### CODE REVIEW

File: src/zarr.cpp [hunk 1]
```diff
@@ -770,6 +778,133 @@ zarr::Zarr::recover_threads_()
...
+                { "clevel", params.clevel_ },
+                { "cname", params.codec_id_ },
+                { "shuffle", params.shuffle_ } } },
+        };
+    }
+
+    auto path = (fs::path(data_root_) / "meta" / "root" /
+                 (std::to_string(level) + ".array.json"))
+                  .string();
+    write_string(path, metadata.dump(4));
```
Comment (andy-sweet) [inline src/zarr.cpp:850]: Optional: not really part of this PR, but I think we should consider passing through the `fs::path` object instead of converting to a `string` here and elsewhere because the first thing we do in `write_string` is convert it back to a path.

File: src/zarr.cpp [hunk 2]
```diff
@@ -770,6 +778,133 @@ zarr::Zarr::recover_threads_()
...
+/// @details Zarr v3 stores group metadata in
+/// /meta/{group_name}.group.json. We will call the group "root".
+void
+zarr::ZarrV3::write_group_metadata_() const
+{
+    namespace fs = std::filesystem;
+    using json = nlohmann::json;
+
+    json metadata;
+    metadata["attributes"]["acquire"] = json::parse(external_metadata_json_);
```
Comment (andy-sweet) [inline src/zarr.cpp:890]: Is it part of the Zarr v3 standard that it's expected to store extra/external metadata in the group attributes (i.e. anyone can put anything they want here)? Or are we assuming this should be a safe enough namespace?

File: src/zarr.hh [hunk 3]
```diff
@@ -132,6 +134,27 @@ struct Zarr final : StorageInterface
     void recover_threads_();
 };
 
+struct ZarrV3 final : public Zarr
```
Comment (andy-sweet) [inline src/zarr.hh:137]: Optional: don't feel strongly enough to really suggest this, but I wonder if we should really use inheritance here. Seems like it's mostly being used to share code rather storing heterogenous collections of zarr and zarrv3 objects and I generally feel like plain old functions are typically better for sharing code.

**Review by andy-sweet — COMMENTED** [review]
Took a first pass through and I don't see any big issues. Will take another pass before approving/requesting changes.

File: src/zarr.cpp (hunk 2 above)
Comment (aliddell) [reply src/zarr.cpp:890]: Yeah, that's my understanding from [this](https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#attributes).

**Review by andy-sweet — APPROVED** [review]
After a second pass, looks good to me. Any remaining comments are optional suggestions, some of which may be better handled outside of this PR.

File: src/zarr.cpp (hunk 1 above)
Comment (aliddell) [reply src/zarr.cpp:850]: No, I think this is a fine idea.
